kernel/src/transaction.rs

```diff
@@ -0,0 +1,142 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+
+use crate::actions::visitors::TransactionVisitor as TxnActionVisitor;
+use crate::actions::{get_log_schema, TRANSACTION_NAME};
+use crate::engine_data::TypedGetData;
+use crate::snapshot::Snapshot;
+use crate::EngineInterface;
+use crate::{actions::Transaction, engine_data::GetData, DataVisitor, DeltaResult};
+
+pub type TransactionMap = HashMap<String, Transaction>;
+pub struct TransactionScanner<'a> {
+    snapshot: Arc<Snapshot>,
+    engine: &'a dyn EngineInterface,
```

Comment:
in general we prefer to not store the engine interface in the struct and to just pass it in where needed. this simplifies the ffi as well as making it so you don't need a lifetime on this struct.